nautobot/extras/models/customfields.py

```diff
@@ -154,20 +151,19 @@ class Meta:
     def __str__(self):
         return self.label or self.name.replace("_", " ").capitalize()
 
-    def remove_stale_data(self, content_types):
-        """
-        Delete custom field data which is no longer relevant (either because the CustomField is
-        no longer assigned to a model, or because it has been deleted).
-        """
-        for ct in content_types:
-            model = ct.model_class()
-            for obj in model.objects.filter(**{f"custom_field_data__{self.name}__isnull": False}):
-                del obj.custom_field_data[self.name]
-                obj.save()
-
     def clean(self):
         super().clean()
 
+        if self.present_in_database:
+            # Check immutable fields
+            database_object = self.__class__.objects.get(pk=self.pk)
+
+            if self.name != database_object.name:
+                raise ValidationError({"name": "Name cannot be changed once created"})
+
+            if self.type != database_object.type:
+                raise ValidationError({"type": "Type cannot be changed once created"})
+
         # Validate the field's default value (if any)
         if self.default is not None:
             try:
```

Comment:
So right now it is a two-step process. Create the field, create the choices, and then update the field with the default. I discussed yesterday that we could work around this with a hacky solution that involves moving this validation to the view. I don't like it for obvious reasons, and it would not be consistent with the API.

Reply:
Thanks for jogging my memory. :-) I can make it a two-step process in the importer as well since we already do two passes of synchronization anyway for other reasons.